File: leetcode-cn/33.搜索旋转排序数组.py
#
# @lc app=leetcode.cn id=33 lang=python3
#
# [33] 搜索旋转排序数组
#

# @lc code=start
from typing import List


class Solution:
    def binarySearch(self, nums, target):
        i = 0
        j = len(nums) - 1

        while i <= j:
            mid = (i + j) // 2

            if nums[mid] < target:
                i += 1
            elif nums[mid] > target:
                j -= 1
            else:
                return mid

        return -1

    def search(self, nums: List[int], target: int) -> int:
        i = 0
        while i < len(nums) - 1 and nums[i] < nums[i + 1]:
            i += 1

        left_res = self.binarySearch(nums[:i + 1], target)

        if left_res != -1:
            return left_res

        right_res = self.binarySearch(nums[i + 1:], target)
        if right_res != -1:
            return i + 1 + right_res
        else:
            return -1


# @lc code=end

File: leetcode-cn/C++/33.搜索旋转排序数组.cpp
/*
 * @lc app=leetcode.cn id=33 lang=cpp
 *
 * [33] 搜索旋转排序数组
 */

// @lc code=start
#include <vector>
using namespace std;
class Solution
{
public:
    int search(vector<int> &nums, int target)
    {
        int n = nums.size();
        int left = 0;
        int right = n - 1;

        while (left <= right)
        {
            int mid = (left + right) / 2;

            if (nums[mid] == target)
            {
                return mid;
            }
            else if (nums[mid] >= nums[0])
            {
                if (target >= nums[0] && target < nums[mid])
                {
                    right = mid - 1;
                }
                else
                {
                    left = mid + 1;
                }
            }
            else
            {
                if (target > nums[mid] && target <= nums[right])
                {
                    left = mid + 1;
                }
                else
                {
                    right = mid - 1;
                }
            }
        }

        return -1;
    }
};
// @lc code=end
